File: src/services/storage.service.ts
// src/services/storage.service.ts
// Persistent storage using react-native-mmkv (fast synchronous key-value store)

import { createMMKV } from 'react-native-mmkv';

const storage = createMMKV({ id: 'meshtalk' });

export const StorageKeys = {
  // Onboarding / Profile
  IS_ONBOARDED: 'is_onboarded',
  NODE_ID: 'node_id',
  NICKNAME: 'nickname',
  PUBLIC_KEY: 'public_key',
  SECRET_KEY: 'secret_key',  // encrypted at rest (ideally use Keychain)

  // BLE
  CONNECTED_DEVICE_ID: 'connected_device_id',
  CONNECTED_DEVICE_NAME: 'connected_device_name',

  // Settings
  THEME: 'theme',               // 'light' | 'dark' | 'system'
  LANGUAGE: 'language',          // 'ru' | 'en'

  // Friends
  FRIENDS: 'friends',            // JSON array

  // Chats
  CHATS: 'chats',                // JSON array of chat metadata
  MESSAGES_PREFIX: 'msgs_',      // msgs_<chatId> = JSON array of messages
} as const;

class StorageService {
  // ─── Generic ───────────────────────────────────────────────────────────────

  getString(key: string): string | undefined {
    return storage.getString(key);
  }

  setString(key: string, value: string): void {
    storage.set(key, value);
  }

  getBool(key: string): boolean | undefined {
    return storage.getBoolean(key);
  }

  setBool(key: string, value: boolean): void {
    storage.set(key, value);
  }

  getJSON<T>(key: string): T | null {
    const raw = storage.getString(key);
    if (!raw) return null;
    try {
      return JSON.parse(raw) as T;
    } catch {
      return null;
    }
  }

  setJSON<T>(key: string, value: T): void {
    storage.set(key, JSON.stringify(value));
  }

  delete(key: string): void {
    storage.remove(key);
  }

  clearAll(): void {
    storage.clearAll();
  }

  // ─── Profile ───────────────────────────────────────────────────────────────

  isOnboarded(): boolean {
    return storage.getBoolean(StorageKeys.IS_ONBOARDED) ?? false;
  }

  setOnboarded(value: boolean): void {
    storage.set(StorageKeys.IS_ONBOARDED, value);
  }

  getNickname(): string {
    return storage.getString(StorageKeys.NICKNAME) ?? '';
  }

  setNickname(nickname: string): void {
    storage.set(StorageKeys.NICKNAME, nickname);
  }

  getNodeId(): string {
    return storage.getString(StorageKeys.NODE_ID) ?? '';
  }

  setNodeId(nodeId: string): void {
    storage.set(StorageKeys.NODE_ID, nodeId);
  }

  getPublicKey(): string {
    return storage.getString(StorageKeys.PUBLIC_KEY) ?? '';
  }

  setPublicKey(key: string): void {
    storage.set(StorageKeys.PUBLIC_KEY, key);
  }

  getSecretKey(): string {
    return storage.getString(StorageKeys.SECRET_KEY) ?? '';
  }

  setSecretKey(key: string): void {
    // In production: use react-native-keychain for secret key storage
    storage.set(StorageKeys.SECRET_KEY, key);
  }

  // ─── Settings ──────────────────────────────────────────────────────────────

  getTheme(): 'light' | 'dark' | 'system' {
    return (storage.getString(StorageKeys.THEME) as any) ?? 'system';
  }

  setTheme(theme: 'light' | 'dark' | 'system'): void {
    storage.set(StorageKeys.THEME, theme);
  }

  getLanguage(): 'ru' | 'en' {
    return (storage.getString(StorageKeys.LANGUAGE) as any) ?? 'ru';
  }

  setLanguage(lang: 'ru' | 'en'): void {
    storage.set(StorageKeys.LANGUAGE, lang);
  }

  // ─── BLE ───────────────────────────────────────────────────────────────────

  getConnectedDeviceId(): string | undefined {
    return storage.getString(StorageKeys.CONNECTED_DEVICE_ID);
  }

  setConnectedDeviceId(id: string | null): void {
    if (id) {
      storage.set(StorageKeys.CONNECTED_DEVICE_ID, id);
    } else {
      storage.remove(StorageKeys.CONNECTED_DEVICE_ID);
    }
  }

  getConnectedDeviceName(): string | undefined {
    return storage.getString(StorageKeys.CONNECTED_DEVICE_NAME);
  }

  setConnectedDeviceName(name: string | null): void {
    if (name) {
      storage.set(StorageKeys.CONNECTED_DEVICE_NAME, name);
    } else {
      storage.remove(StorageKeys.CONNECTED_DEVICE_NAME);
    }
  }

  // ─── Messages ──────────────────────────────────────────────────────────────

  getMessages(chatId: string): StoredMessage[] {
    return this.getJSON<StoredMessage[]>(`${StorageKeys.MESSAGES_PREFIX}${chatId}`) ?? [];
  }

  saveMessage(chatId: string, message: StoredMessage): void {
    const messages = this.getMessages(chatId);
    const idx = messages.findIndex(m => m.id === message.id);
    if (idx >= 0) {
      messages[idx] = message;
    } else {
      messages.push(message);
    }
    // Keep last 500 messages per chat
    const trimmed = messages.slice(-500);
    this.setJSON(`${StorageKeys.MESSAGES_PREFIX}${chatId}`, trimmed);
  }

  updateMessageStatus(chatId: string, messageId: string, status: MessageStatus): void {
    const messages = this.getMessages(chatId);
    const idx = messages.findIndex(m => m.id === messageId);
    if (idx >= 0) {
      messages[idx].status = status;
      this.setJSON(`${StorageKeys.MESSAGES_PREFIX}${chatId}`, messages);
    }
  }

  deleteMessages(chatId: string): void {
    storage.remove(`${StorageKeys.MESSAGES_PREFIX}${chatId}`);
  }
}

// ─── Types ─────────────────────────────────────────────────────────────────

export type MessageStatus = 'sending' | 'sent' | 'delivered' | 'read' | 'failed';

export interface StoredMessage {
  id: string;
  chatId: string;
  senderId: string;       // nodeId
  senderNickname: string;
  text: string;           // decrypted text (stored decrypted locally for speed)
  timestamp: number;
  status: MessageStatus;
  isOutgoing: boolean;
  packetId?: string;      // for ACK matching
}

export const storageService = new StorageService();

File: lib/services/storage_service.dart
import 'package:flutter/material.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:hive_flutter/hive_flutter.dart';

import '../models/chat.dart';
import '../models/friend.dart';
import '../models/message.dart';
import '../models/profile.dart';

class StorageService {
  static const _secure = FlutterSecureStorage();

  static const _boxApp = 'app';
  static const _boxFriends = 'friends';
  static const _boxChats = 'chats';
  static const _boxMessages = 'messages';

  static Future<void> init() async {
    await Hive.initFlutter();
    await Hive.openBox(_boxApp);
    await Hive.openBox(_boxFriends);
    await Hive.openBox(_boxChats);
    await Hive.openBox(_boxMessages);
  }

  static Box get app => Hive.box(_boxApp);
  static Box get friends => Hive.box(_boxFriends);
  static Box get chats => Hive.box(_boxChats);
  static Box get messages => Hive.box(_boxMessages);

  static bool get isOnboarded => (app.get('isOnboarded') as bool?) ?? false;
  static set isOnboarded(bool v) => app.put('isOnboarded', v);

  static String get languageCode => (app.get('language') as String?) ?? 'ru';
  static set languageCode(String v) => app.put('language', v);

  static ThemeMode get themeMode {
    final raw = (app.get('themeMode') as String?) ?? 'system';
    switch (raw) {
      case 'light':
        return ThemeMode.light;
      case 'dark':
        return ThemeMode.dark;
      default:
        return ThemeMode.system;
    }
  }

  static set themeMode(ThemeMode mode) => app.put('themeMode', mode.name);

  static Future<void> savePrivateKey(String key) => _secure.write(key: 'privateKey', value: key);
  static Future<String?> getPrivateKey() => _secure.read(key: 'privateKey');

  static UserProfile? get profile {
    final raw = app.get('profile');
    if (raw is Map) return UserProfile.fromJson(raw);
    return null;
  }

  static Future<void> saveProfile(UserProfile profile) => app.put('profile', profile.toJson());

  static List<FriendModel> getAllFriends() {
    return friends.values
        .whereType<Map>()
        .map(FriendModel.fromJson)
        .toList(growable: false);
  }

  static Future<void> upsertFriend(FriendModel friend) => friends.put(friend.nodeId, friend.toJson());

  static List<ChatModel> getAllChats() {
    return chats.values.whereType<Map>().map(ChatModel.fromJson).toList(growable: false)
      ..sort((a, b) => b.lastMessageAt.compareTo(a.lastMessageAt));
  }

  static Future<void> upsertChat(ChatModel chat) => chats.put(chat.id, chat.toJson());

  static List<MessageModel> getMessages(String chatId) {
    final raw = messages.get(chatId);
    if (raw is! List) return const [];
    return raw.whereType<Map>().map(MessageModel.fromJson).toList(growable: false)
      ..sort((a, b) => a.timestamp.compareTo(b.timestamp));
  }

  static Future<void> saveMessage(String chatId, MessageModel message) async {
    final list = getMessages(chatId).toList();
    final idx = list.indexWhere((m) => m.id == message.id);
    if (idx >= 0) {
      list[idx] = message;
    } else {
      list.add(message);
    }
    await messages.put(chatId, list.map((e) => e.toJson()).toList());
  }
}
